### PR TITLE
CRAYSAT-1319: Add report output for `sat bootprep` results

### PR DESCRIPTION
**Pull request by jack-stanek-hpe**
## Summary and Scope

This change adds a reporting capability to `sat bootprep` which reports
which items (configs, images, and session templates) were created during
a sat bootprep run. This information is printed as a report and its
format can be set to 'pretty', 'yaml' or 'json' based on the context of
the run. (For instance, JSON output is useful for IUF runs, where a
pretty report is more useful for end-user runs.) Unit tests for the new 
functionality in the BaseInputItem class were also added.

This change also refactors the Report class to make it easier to dump JSON
and YAML reports. This change also adds a new MultiReport class which
acts as a container for multiple Reports such that they have consistent
formatting and display options. This can also be reused for other
applications where multiple reports are printed such as in `sat
showrev`.

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CRAYSAT-1319](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1319)

### Tested on:

  * `frigg`

### Test description:

Run unit tests. Manually test `sat bootprep` against a minimal input file. Test that ouptut is as expected.

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: tests/test_report.py [hunk 1]
```diff
@@ -677,3 +677,66 @@ def test_show_empty_show_missing(self):
...
+
+    def test_pretty_printing_multireport(self):
+        """Test pretty-printing a MultiReport"""
+        pretty_report = self.multireport.get_formatted_report('pretty')
+        self.assertIn(self.recipe_report.get_formatted_report('pretty'), pretty_report)
+        self.assertIn(self.variables_report.get_formatted_report('pretty'), pretty_report)
+
+    def test_multireport_with_multiple_formats(self):
+        """Test that a MultiReport with multiple formats in use defaults to pretty format"""
+        for recipe_format, variables_format in combinations(['pretty', 'json', 'yaml'], 2):
```
Comment (haasken-hpe) [inline tests/test_report.py:718]: I think you might have meant to use `combinations_with_replacement` if you are expecting to test the case when they are actually equal below.

**Review by haasken-hpe — COMMENTED** [review]
Of all the comments, the one I feel is most important to address is the one about a possible incorrectly handled edge case in the reporting of the created images.

**Comment by jack-stanek-hpe** [comment]
Test output with image creation here: https://gist.github.com/jack-stanek-hpe/e709ff9f068141c66440f2de10d0698e

File: sat/cli/bootprep/image.py [hunk 2]
```diff
@@ -490,3 +491,7 @@ def create_images(instance, args):
     if args.delete_ims_jobs:
         for image in images_to_create:
             image.clean_up_image_create_job()
+
+    # TODO: This is pretty ugly, and would probably be better for it to go in
+    #  the GroupWaiter class or the future InputImageCollection class.
+    return set(waiter.members) - set(waiter.pending | waiter.failed)
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply sat/cli/bootprep/image.py:497]: I tested this case, and it looks like we don't ever really get to this point of returning the images that were created if there is a failure creating images. Here's how I tested: https://gist.github.com/haasken-hpe/f15f20208da2f3dd178a9cb38dccb2d1

Basically I ran bootprep against a file that created two images, one of which depended on the other. Then I deleted the IMS image creation job for the first image to get it to fail waiting on it.

[the remark this answers is not in the document]
Comment (haasken-hpe) [reply sat/cli/bootprep/image.py:497]: It would probably be good to improve this like you said in the comment, but since any image creation failure causes us to raise `ImageCreateError` instead of returning output here, it's probably fine as-is for IUF. I think in the case that all images are created successfully, this set difference should be correct.

File: sat/cli/bootprep/input/base.py [hunk 3]
```diff
@@ -518,6 +532,7 @@ def create_items(self):
                     failed_items.append(item)
                     LOGGER.error(f'Failed to create {item_description}: {err}')
                 else:
+                    self.created.append(item)
```
[the remark this answers is not in the document]
Comment (jack-stanek-hpe) [reply sat/cli/bootprep/input/base.py:535]: I can add the skipped images too, that shouldn't be too hard, though the schema of the report might be different. To make this work with images too, some additional work might be needed to add that functionality in addition to how I'm already handling them specially.

File: sat/report.py [hunk 4]
```diff
@@ -359,6 +382,17 @@ def get_pretty_table(self):
 
         return pt
 
+    def get_dumpable_structure(self, include_title=True):
```
[the remark this answers is not in the document]
Comment (jack-stanek-hpe) [reply sat/report.py:385]: Good point, I've removed it. I was using it for a previous implementation and it slipped through.

File: sat/report.py [hunk 5]
```diff
@@ -359,6 +382,17 @@ def get_pretty_table(self):
...
+        """Get a structure which can be natively dumped as YAML or JSON."""
+        _, rows_to_print = self.get_rows_to_print()
+
+        if include_title and not self.no_headings and self.title:
+            return {self.title: rows_to_print}
+        else:
+            if not rows_to_print:
+                return ''
```
[the remark this answers is not in the document]
Comment (jack-stanek-hpe) [reply sat/report.py:393]: Having an empty list ought to be fine, actually. I'm not sure what my reasoning for this was.

File: tests/cli/bootprep/input/test_base.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        row_headings = ['name', 'type']
+        row_vals = ['foo', 'configuration']
+        item_cls = create_input_item_cls(row_headings)
+        self.assertEqual(item_cls(dict(zip(row_headings, row_vals))).report_row(),
+                         row_vals)
+
+    def test_report_rows_excluded(self):
+        """Test that report rows are not included if not present in report_attrs"""
+        item_cls = create_input_item_cls(['name'])
+        self.assertEqual(item_cls({'name': 'foo'}).report_row(), ['foo'])
```
[the remark this answers is not in the document]
Comment (jack-stanek-hpe) [reply tests/cli/bootprep/input/test_base.py:75]: Ah, yes, good catch.

File: sat/cli/bootprep/input/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jack-stanek-hpe) [reply sat/cli/bootprep/input/base.py:535]: Created [CRAYSAT-1652](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1652) to encapsulate this work.

File: sat/cli/bootprep/main.py [hunk 7]
```diff
@@ -324,6 +324,7 @@ def print_report(args, instance, created_images):
             item_class = IMSInputImage
         else:
             created = items.created
+            skipped = items.skipped_items
```
Comment (haasken-hpe) [inline sat/cli/bootprep/main.py:327]: Unused variable.

**Review by haasken-hpe — APPROVED** [review]

File: sat/cli/bootprep/main.py (hunk 7 above)
Comment (jack-stanek-hpe) [reply sat/cli/bootprep/main.py:327]: Doh, I was not supposed to have committed that!
